### PR TITLE
meas_base review for DM-4926: centroids outside Footprints.

### PR DESCRIPTION
**Pull request by TallJimbo**

### CODE REVIEW

File: include/lsst/meas/base/CentroidUtilities.h [hunk 1]
```diff
@@ -201,6 +201,42 @@ class CentroidTransform : public BaseTransform {
     afw::table::CovarianceMatrixKey<ErrElement,2> _coordErrKey;
 };
 
+class CentroidChecker {
+
```
Comment (TallJimbo) [inline include/lsst/meas/base/CentroidUtilities.h]: Quibble: this newline is unnecessary.

File: include/lsst/meas/base/CentroidUtilities.h [hunk 2]
```diff
@@ -201,6 +201,42 @@ class CentroidTransform : public BaseTransform {
...
+     *     (3) change the value of the centroid to the footprint Peak
+     *
+     *  Secondly, check to see if the centroid is more than "dist" pixels from
+     *  the footprint peak, and similarly modify the centroi value to the peak in that case
+     *
+     *  @param[out] schema   Schema to which the flag_resetToPeak is to be added
+     *  @param[in]  name     The name of the algorithm we will be checking
+     *  @param[in]  dist     maximum distance from footprint peak for distance check
+     */
+    CentroidChecker(afw::table::Schema & schema, std::string const & name, int dist=-1);
```
Comment (TallJimbo) [inline include/lsst/meas/base/CentroidUtilities.h]: Please make `dist` a `double` (here and elsewhere).

File: include/lsst/meas/base/CentroidUtilities.h [hunk 3]
```diff
@@ -201,6 +201,42 @@ class CentroidTransform : public BaseTransform {
...
+     *  already set by the algorithm is outside the footprint attached to the record.
+     *  If that is the case, we should:
+     *     (1) set the general failure flag for the algorithm
+     *     (2) set the algorithmName + "_flag_resetToPeak" flag
+     *     (3) change the value of the centroid to the footprint Peak
+     *
+     *  Secondly, check to see if the centroid is more than "dist" pixels from
+     *  the footprint peak, and similarly modify the centroi value to the peak in that case
+     *
+     *  @param[out] schema   Schema to which the flag_resetToPeak is to be added
```
Comment (TallJimbo) [inline include/lsst/meas/base/CentroidUtilities.h]: Quibble: this should probably be `@param[in,out] schema ...`

**Comment by TallJimbo** [comment]
It wasn't introduced on this issue, but since I've noticed it here, could you fix the "instnace" type in the `addFlagHandler` docstring?  Separate commit, please.

File: tests/testCentroidChecker.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        self.xKey = schema.addField(schema.join(name, "x"), type=numpy.float64)
+        self.yKey = schema.addField(schema.join(name, "y"), type=numpy.float64)
+        if self.config.dist is None:
+            self.centroidChecker = lsst.meas.base.CentroidChecker(schema, name)
+        else:
+            self.centroidChecker = lsst.meas.base.CentroidChecker(schema, name, self.config.dist)
+
+    #   This is a measure routine which does nothing except to raise Exceptions
+    #   as requested by the caller. Errors normally don't occur unless there is
+    #   something wrong in the inputs, or if there is an error during the measurement
```
Comment (TallJimbo) [inline tests/testCentroidChecker.py]: I find "as requested by the caller" a bit confusing here.  do you mean "Exceptions caused by moving the true result the amount specified in the config"?

File: tests/testCentroidChecker.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    suites += unittest.makeSuite(lsst.utils.tests.MemoryTestCase)
+    return unittest.TestSuite(suites)
+
+
+def run(shouldExit=False):
+    """Run the tests"""
+    lsst.utils.tests.run(suite(), shouldExit)
+
+if __name__ == "__main__":
+    run(True)
```
Comment (TallJimbo) [inline tests/testCentroidChecker.py]: There's now a new recommended way to do the boilerplate at the end of Python tests.  You can find the instructions in the [SQR-012](https://sqr-012.lsst.io/) tech note.

File: include/lsst/meas/base/CentroidUtilities.h (hunk 2 above)
Comment (pgee2000) [inline include/lsst/meas/base/CentroidUtilities.h]: Ok, done.

File: include/lsst/meas/base/CentroidUtilities.h (hunk 3 above)
Comment (pgee2000) [inline include/lsst/meas/base/CentroidUtilities.h]: OK

File: tests/testCentroidChecker.py (hunk 4 above)
Comment (pgee2000) [inline tests/testCentroidChecker.py]: This comment is no longer relevant.  It was for a previous incarnation.

**Pull request merged**